### PR TITLE
[JIT] Fix `@staticmethod` access from `self` on modules

### LINKED ISSUE #30755: [jit] `@staticmethod`s retrieved from `self` don't work
```python
class M(nn.Module):
    @staticmethod
    def some_method(x):
        return x + 10

    def forward(self, x):
        return self.some_method(x)

torch.jit.script(M())
```

The compiler treats `some_method` as a regular method instead of a static method, so `x` ends up having the type `M` instead of `Tensor`.

The error is: 
```
RuntimeError:
Tried to access nonexistent attribute or method '__add__' of type '__torch__.M'. Did you forget to initialize an attribute in __init__()?:
  File "/private/home/suo/scratch/test.py", line 21
    @staticmethod
    def some_method(x):
        return x + 10
               ~~~~~~ <--- HERE
'M.some_method' is being compiled since it was called from 'M.forward'
  File "/private/home/suo/scratch/test.py", line 24
    def forward(self, x):
        return self.some_method(x)
               ~~~~~~~~~~~~~~~~~~ <--- HERE
```
cc @suo

### PR DESCRIPTION
**Pull request by voznesenskym**
Closes #30755

### CODE REVIEW

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 1]
```diff
@@ -753,7 +764,7 @@ std::shared_ptr<SugaredValue> toSugaredValue(
     }
   }
 
-  py::bool_ isFunction = py::module::import("inspect").attr("isfunction")(obj);
+  py::bool_ isFunction = py::module::import("inspect").attr("isroutine")(obj);
```
Comment (suo) [inline torch/csrc/jit/python/python_sugared_value.cpp:767]: Can you explain why this was changed? I ran the following script: 
```
class Foo:
    @staticmethod
    def my_static(x, y):
        return x + y
from inspect import getattr_static

def is_static_fn(cls, fn):
    return isinstance(getattr_static(cls, fn), staticmethod)

def get_static_fn(cls, fn):
    return getattr_static(cls, fn).__func__

print(inspect.isfunction(get_static_fn(Foo, "my_static")))
> True
```
So it seems like we don't need `isroutine` to detect the static functions returned by the new helpers this PR adds. From the docs, it seems the only difference is that it includes builtins, which doesn't seem like it's that relevant?

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 2]
```diff
@@ -44,6 +44,7 @@ FunctionSchema PythonValue::getSchema(
   auto py_param_names = annotations.attr("get_param_names")(callable, n_args);
   auto param_names = py::cast<std::vector<std::string>>(py_param_names);
   auto names_it = param_names.begin();
+
```
Comment (suo) [inline torch/csrc/jit/python/python_sugared_value.cpp:47]: nit: some empty lines added

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 3]
```diff
@@ -426,7 +427,17 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
...
   if (py::isinstance<py::function>(unboundMethod)) {
+    bool isStaticFn =
+        py::cast<bool>(py::module::import("torch._jit_internal")
+                           .attr("is_static_fn")(concreteType_->getPyClass(), field.c_str()));
+    if (isStaticFn) {
+      // Functions within the module annotated with @staticmethod do not need binding.
+      py::object staticFn = py::module::import("torch._jit_internal")
+                           .attr("get_static_fn")(concreteType_->getPyClass(), field.c_str());
+      return toSugaredValue(staticFn, m, loc);
```
Comment (suo) [inline torch/csrc/jit/python/python_sugared_value.cpp:439]: nice!

File: torch/_jit_internal.py [hunk 4]
```diff
@@ -453,6 +453,15 @@ def is_ignored_fn(fn):
     mod = get_torchscript_modifier(fn)
     return mod is FunctionModifiers.UNUSED or mod is FunctionModifiers.IGNORE
 
+from inspect import getattr_static
```
Comment (suo) [inline torch/_jit_internal.py:456]: let's put this at the top

**Review by suo — COMMENTED** [review]
Nice! General approach looks solid, had a question inline about the implementation.

Also, your commit accidentally added some submodule changes (`third_party/fbgemm`).

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 1 above)
Comment (voznesenskym) [reply torch/csrc/jit/python/python_sugared_value.cpp:767]: This is indeed not needed. It seemed to throw errors before, but it was probably something else and I got my wires crossed. Removed.

**Review by suo — APPROVED** [review]
lgtm!

**Pull request closed without merging**